store/structured/mongo/store.go

```diff
@@ -0,0 +1,184 @@
+package mongo
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	mongoDriver "go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+
+	"github.com/tidepool-org/platform/errors"
+)
+
+type Store struct {
+	client          *mongoDriver.Client
+	config          *Config
+	initializeGroup sync.WaitGroup
+	closingChannel  chan bool
+	clientMux       sync.Mutex
+}
+
+type Status struct {
+	Ping string
+}
+
+func (s *Store) getClient() *mongoDriver.Client {
+	s.clientMux.Lock()
+	defer s.clientMux.Unlock()
+	return s.client
+}
+func (s *Store) setClient(cli *mongoDriver.Client) {
+	s.clientMux.Lock()
+	defer s.clientMux.Unlock()
+	s.client = cli
+}
+
+func (s *Store) Close() error {
+	if s.closingChannel != nil {
+		s.closingChannel <- true
+	}
+	s.initializeGroup.Wait()
+	return s.getClient().Disconnect(context.Background())
+}
+
+func NewStore(c *Config) (*Store, error) {
+	if c == nil {
+		return nil, errors.New("database config is empty")
+	} else if err := c.Validate(); err != nil {
+		return nil, errors.Wrap(err, "config is invalid")
+	}
+
+	store := &Store{
+		config: c,
+	}
+	store.Start()
+	return store, nil
+}
+
+func (s *Store) Start() {
+	if s.closingChannel == nil {
+		s.initializeGroup.Add(1)
+		go s.connectionRoutine()
+	}
+}
+
+func (s *Store) connectionRoutine() {
+
+	var attempts int64 = 1
+	var err error
+	s.closingChannel = make(chan bool, 1)
+	for {
+		var timer <-chan time.Time
+		if attempts == int64(0) {
+			timer = time.After(0)
+		} else {
+			timer = time.After(s.config.WaitConnectionInterval)
+		}
+		select {
+		case <-s.closingChannel:
+			close(s.closingChannel)
+			s.closingChannel = nil
+			s.initializeGroup.Done()
+			return
+		case <-timer:
+			err = initConnexion(s)
+			if err == nil {
+				s.closingChannel <- true
+			} else {
+				if s.config.MaxConnectionAttempts > 0 && s.config.MaxConnectionAttempts <= attempts {
+					s.closingChannel <- true
```

Comment:
We should crash the service in that case, or it will remain up without any db connection. 
```suggestion
					s.closingChannel <- true
					panic(err)
```